source/adapters/level_zero/v2/latency_tracker.hpp

```diff
@@ -0,0 +1,212 @@
+};
+
+inline latency_printer &globalPrinter() {
+  static latency_printer printer;
+  return printer;
+}
+
+class latency_histogram {
+public:
+  inline latency_histogram(const char *name,
+                           latency_printer &printer = globalPrinter(),
+                           int64_t lowestDiscernibleValue = 1,
+                           int64_t highestTrackableValue = 100'000'000'000,
+                           int significantFigures = 3)
+      : name(name), histogram(nullptr, nullptr), printer(printer) {
+    if (trackLatency) {
+      struct hdr_histogram *cHistogram;
+      auto ret = hdr_init(lowestDiscernibleValue, highestTrackableValue,
+                          significantFigures, &cHistogram);
+      if (ret != 0) {
+        logger::error("Failed to initialize latency histogram");
+      }
+      histogram = std::unique_ptr<struct hdr_histogram, decltype(&hdr_close)>(
+          cHistogram, &hdr_close);
+    }
+  }
+
+  latency_histogram(const latency_histogram &) = delete;
+  latency_histogram(latency_histogram &&) = delete;
+
+  inline ~latency_histogram() {
+    if (!trackLatency || !histogram) {
+      return;
+    }
+
+    if (hdr_min(histogram.get()) == std::numeric_limits<int64_t>::max()) {
+      logger::info("[{}] latency: no data", name);
+      return;
+    }
+
+    printer.publishLatency(name, std::move(histogram));
+  }
+
+  inline void trackValue(int64_t value) {
+    hdr_record_value(histogram.get(), value);
+  }
+
+private:
+  const char *name;
+  histogram_ptr histogram;
+  latency_printer &printer;
+};
+
+class latency_tracker {
+public:
+  inline explicit latency_tracker(latency_histogram &stats)
+      : stats(trackLatency ? &stats : nullptr), begin() {
+    if (trackLatency) {
+      begin = std::chrono::steady_clock::now();
+    }
+  }
+  inline latency_tracker() {}
+  inline ~latency_tracker() {
+    if (stats) {
+      auto tp = std::chrono::steady_clock::now();
+      auto diffNanos =
+          std::chrono::duration_cast<std::chrono::nanoseconds>(tp - begin)
+              .count();
+      stats->trackValue(static_cast<int64_t>(diffNanos));
+    }
+  }
+
+  latency_tracker(const latency_tracker &) = delete;
+  latency_tracker &operator=(const latency_tracker &) = delete;
+
+  inline latency_tracker(latency_tracker &&rhs) noexcept
+      : stats(rhs.stats), begin(rhs.begin) {
+    rhs.stats = nullptr;
+  }
+
+  inline latency_tracker &operator=(latency_tracker &&rhs) noexcept {
+    if (this != &rhs) {
+      this->~latency_tracker();
+      new (this) latency_tracker(std::move(rhs));
+    }
+    return *this;
+  }
+
+private:
+  latency_histogram *stats{nullptr};
+  std::chrono::time_point<std::chrono::steady_clock> begin;
+};
+
+// To resolve __COUNTER__
+#define CONCAT(a, b) a##b
+
+// Each tracker has it's own thread-local histogram.
+// At program exit, all histograms for the same scope are
+// aggregated.
+#define TRACK_SCOPE_LATENCY_CNT(name, cnt)                                     \
```

Comment:
just a suggestion: you could make the name optional and use `__FUNCTION__`.

Reply:
Yeah, I thought about this, but when you're using this on member methods, __FUNCTION__ doesn't tell you the class name (and we might have the same methods' names) so I figures it would be better to just pass a name always

Reply:
`__PRETTY_FUNCTION__` ? Google says it should include the class name, but we'd need a workaround for visual studio:
```
#if !defined(__PRETTY_FUNCTION__) && !defined(__GNUC__)
#define __PRETTY_FUNCTION__ __FUNCSIG__
#endif
```

Reply:
Pretty function on the other hand prints too much imho ;d It includes the return type and all the arguments, so instead of `command_list_cache_t::getImmediateCommandList` we would have `raii::cache_borrowed_command_list_t
command_list_cache_t::getImmediateCommandList(
    ze_device_handle_t, bool, uint32_t,
    ze_command_queue_mode_t, ze_command_queue_priority_t,
    std::optional<uint32_t>)`

Reply:
makes sense.